source/blog/2017-10-10-glimmer-progress-report.md

```diff
@@ -0,0 +1,645 @@
+equivalent compiled JavaScript. We received many reports of apps dropping 30-50%
+in total (post-gzip!) size after upgrading to Ember 2.10, the first version to
+use this JSON-based approach.
+
+As exciting as this was, we knew that JSON was not the final word in compactly
+and efficiently representing compiled templates.
+
+At runtime, Glimmer VM would gather the JSON for each template and compile them
+into a final, internal representation that was just a large array of 32-bit
+integers. After looking at traces of real-world Glimmer.js apps, we knew we
+could improve boot times by precomputing this final compilation step at build
+time.
+
+Helpfully, browsers have become increasingly fluent at dealing with binary data,
+largely driven by demanding multimedia use cases like audio, video, and 3D
+graphics. And while JSON is fast to parse, as the old saw goes, no parse is faster
+than *no* parse. What if we could serialize compiled templates into a binary format
+that the VM could start executing _without a parse step_?
+
+I'm no M. Night Shyamalan, so you've probably already guessed the ending here:
+that's exactly what we've done. Recent versions of Glimmer VM include the
+`@glimmer/bundle-compiler` package, our name for the compiler that produces a
+binary "bundle" of all of your compiled templates.
+
+We are planning to land support for binary templates as an opt-in in Glimmer.js
+soon. (The feature is already landed in the low-level Glimmer VM but is not yet
+exposed in a convenient way.)
+
+One thing to note about the bundle compiler is that it requires knowing your
+entire program statically at build time. The browser tends to be a pretty
+dynamic environment, however, so Glimmer VM still supports "lazy compilation"
+(i.e. compiling to JSON) as a first-class mode.
+
+In the Ember ecosystem, apps and addons do very dynamic things (like register
+components at runtime) which are incompatible with the bundle compiler. We want
+to enable binary templates in Ember, but this is farther out because we will
+need to figure out exactly what the constraints are and provide guidance for
+app and addon authors.
+
+In exchange for the (admittedly pretty incredible) performance benefits, binary
+templates also introduce extra complexity.
+
+Binary templates can't be inlined in HTML or JavaScript, so they must be fetched
+as early as possible in the page lifecycle. No browser I tested yet supports
+`<link rel="preload" as="fetch">`, which would allow a streaming HTML parser to
+detect and fetch binary data very early in the page load. No tools or CDNs know
+what the heck a `.gbx` file is (the file extension of compiled binaries), and
+require manual configuration. You probably want H2 Push for this, but that's
+its own can of worms.
+
+Getting these optimally deployed will probably suck for a while, but I have
+faith that the Ember community will do what it does best and rally around a set
+of shared, high quality solutions for dealing with this.
```

Comment:
Some where in this section (not necessarily at the end) you should link to Chad's write-up IMO: https://www.linkedin.com/pulse/glimmers-optimizing-compiler-chad-hietala/